win32ss/user/ntuser/ghost.c

```diff
@@ -1,18 +1,175 @@
+        if (Len > 0)
+        {
+            Ret = RtlEqualUnicodeString(&ClassName, &GhostClass, TRUE);
+        }
+        else
+        {
+            DPRINT1("Unable to get class name\n");
+        }
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        DPRINT1("Exception!\n");
+        Ret = FALSE;
+    }
+    _SEH2_END;
+
+    RtlFreeUnicodeString(&ClassName);
+
+    return Ret;
+}
+
+HWND APIENTRY NtUserGhostWindowFromHungWindow(HWND hwndHung)
+{
+    RTL_ATOM Atom;
+    PWND pHungWnd;
+    PPROPERTY Prop;
+    HWND hwndGhost;
+
+    pHungWnd = ValidateHwndNoErr(hwndHung);
+    if (!pHungWnd)
+    {
+        DPRINT("Not a window\n");
+        return NULL;
+    }
+
+    if (!IntGetAtomFromStringOrAtom(&GhostProp, &Atom))
+        ASSERT(FALSE);
+
+    Prop = UserGetProp(pHungWnd, Atom, TRUE);
+    if (Prop)
+    {
+        hwndGhost = (HWND)Prop->Data;
+        if (ValidateHwndNoErr(hwndGhost))
+            return hwndGhost;
+
+        DPRINT("Not a window\n");
+    }
+
+    return NULL;
+}
+
+HWND APIENTRY NtUserHungWindowFromGhostWindow(HWND hwndGhost)
+{
+    const GHOST_DATA *UserData;
+    PWND pGhostWnd;
+    HWND hwndTarget;
+
+    pGhostWnd = ValidateHwndNoErr(hwndGhost);
+    if (!IntIsGhostWindow(pGhostWnd))
+    {
+        DPRINT("Not a ghost window\n");
+        return NULL;
+    }
+
+    UserData = (const GHOST_DATA *)pGhostWnd->dwUserData;
+    if (UserData)
+    {
+        _SEH2_TRY
+        {
+            ProbeForRead(UserData, sizeof(GHOST_DATA), 1);
+            hwndTarget = UserData->hwndTarget;
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+        {
+            DPRINT1("Exception!\n");
+            hwndTarget = NULL;
+        }
+        _SEH2_END;
+    }
+    else
+    {
+        DPRINT("No user data\n");
+        hwndTarget = NULL;
+    }
+
+    if (hwndTarget)
+    {
+        if (ValidateHwndNoErr(hwndTarget))
+            return hwndTarget;
+
+        DPRINT1("Not a window\n");
+    }
+
+    return NULL;
+}
+
+BOOL FASTCALL IntMakeHungWindowGhosted(HWND hwndHung)
+{
+    PWND pHungWnd = ValidateHwndNoErr(hwndHung);
+    if (!pHungWnd)
+    {
+        DPRINT1("Not a window\n");
+        return FALSE;   // not a window
+    }
+
+    if (IntIsGhostWindow(pHungWnd))
+    {
+        DPRINT1("IntIsGhostWindow\n");
+        return FALSE;   // ghost window cannot be ghosted
+    }
+
+    if (!MsqIsHung(pHungWnd->head.pti))
+    {
+        DPRINT1("Not hung window\n");
+        return FALSE;   // not hung window
+    }
+
+    if (!(pHungWnd->style & WS_VISIBLE))
+        return FALSE;   // invisible
+
+    if (pHungWnd->style & WS_CHILD)
+        return FALSE;   // child
+
+    if (NtUserGhostWindowFromHungWindow(hwndHung))
+    {
+        DPRINT("Already ghosting\n");
+        return FALSE;   // already ghosting
+    }
+
     // TODO:
     // 1. Create a thread.
     // 2. Create a ghost window in the thread.
     // 3. Do message loop in the thread
-    STUB;
+
     return FALSE;
```

Comment:
So this function basically always returns FALSE, am I right?

Reply:
Right. That's TODO.